### PR TITLE
Fix/libpq mingw static

### LINKED ISSUE #11716: [package] libpq/*: static MinGW build
### Package and Environment Details

* Package Name/Version: **libpq/\***
* Operating System+version: **Windows 10**
* Compiler+version: **MinGW/GCC**

Commit 2f3a500 introduces a check that is supposed to prevent static builds in MinGW (later moved to the `validate()` method). But the condition for this check here tests for `options.shared == True`, so it raises an exception for building a shared library. As I understand, the opposite should be the case.

For curiosity: What is actually the reason for this limitation? I couldn't find any explanation for this in the internet.

### LINKED ISSUE #15301: [package] libpq/14.5: app linked against libpq depends libpq.dll despite option libpq:shared=False
### Description

I built a test app which uses libpq from conan, explicitly setting `libpq:shared=False`
Nevertheless, the generated exe requires libpq.dll to run. I would expext libpq to be built and linked statically when the option `shared` is set to `False`

Here is a minamal test project, which demonstrates the problem:

#### conanfile.txt:
```
[build_requires]

[requires]
libpq/14.5

[options]
libpq:shared=False

[generators]
virtualenv
CMakeDeps
CMakeToolchain
```

####  CMakeLists.txt
```
cmake_minimum_required(VERSION 3.24)
project(pqtest)
find_package(PostgreSQL)
add_executable(pqtest pqtest.cpp)
target_link_libraries(pqtest PRIVATE PostgreSQL::PostgreSQL)
```

#### pqtest.cpp
```
#include <libpq-fe.h>
#include <iostream>

int main()
{
    int lib_ver = PQlibVersion();
    std::cout << lib_ver << "\n";
    return 0;
}
```

### Package and Environment Details

* Package Name/Version: **libpq/14.5**
* Operating System+version: **Windows 11**
* Compiler+version: **GCC 12.2 (from mingw-builds/12.2.0)**
* Docker image: **no**
* Conan version: **conan 1.57.0**
* Python version: **Python 3.11.0**

### Conan profile

[settings]
os=Windows
arch=x86_64
build_type=Release
compiler=gcc
compiler.threads=posix
compiler.libcxx=libstdc++11
compiler.version=12
compiler.exception=seh
[options]
[conf]
[build_requires]
*: mingw-builds/12.2.0
[env]

### Steps to reproduce

`conan install .. --build=missing -pr:h=mingw -pr:b=mingw`
`activate.bat`
`cmake .. -G "MinGW Makefiles" -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release`
`mingw32-make`

### Logs

<details><summary>Click to expand log</summary>

```
C:\Proj\pqtest>mkdir build

C:\Proj\pqtest>cd build

C:\Proj\pqtest\build>conan install .. --build=missing -pr:h=mingw -pr:b=mingw
Configuration (profile_host):
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.exception=seh
compiler.libcxx=libstdc++11
compiler.threads=posix
compiler.version=12
os=Windows
[options]
[build_requires]
*: mingw-builds/12.2.0
[env]

Configuration (profile_build):
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.exception=seh
compiler.libcxx=libstdc++11
compiler.threads=posix
compiler.version=12
os=Windows
[options]
[build_requires]
*: mingw-builds/12.2.0
[env]

conanfile.txt: Installing package
Requirements
    libpq/14.5 from 'conancenter' - Cache
Packages
    libpq/14.5:72ab1929946a1e7e9b69c00bc299eb0126492ede - Cache
Build requirements
    mingw-builds/12.2.0 from 'conancenter' - Cache
Build requirements packages
    mingw-builds/12.2.0:6903a9d1b48b06f7fbe0929db654512c77e6cc32 - Cache

Installing (downloading, building) binaries...
libpq/14.5: Already installed!
mingw-builds/12.2.0: Already installed!
mingw-builds/12.2.0: Appending PATH env var with : C:\Users\rainer\.conan\data\mingw-builds\12.2.0\_\_\package\6903a9d1b48b06f7fbe0929db654512c77e6cc32\bin
conanfile.txt: Applying build-requirement: mingw-builds/12.2.0
conanfile.txt: Generator txt created conanbuildinfo.txt
conanfile.txt: Generator 'CMakeToolchain' calling 'generate()'
conanfile.txt: Preset 'release' added to CMakePresets.json. Invoke it manually using 'cmake --preset release'
conanfile.txt: If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "MinGW Makefiles" -DCMAKE_TOOLCHAIN_FILE=C:\Proj\pqtest\build\conan_toolchain.cmake -DCMAKE_SH=CMAKE_SH-NOTFOUND -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
conanfile.txt: Generator virtualenv created activate.bat
conanfile.txt: Generator virtualenv created deactivate.bat
conanfile.txt: Generator virtualenv created environment.bat.env
conanfile.txt: Generator virtualenv created activate.ps1
conanfile.txt: Generator virtualenv created deactivate.ps1
conanfile.txt: Generator virtualenv created environment.ps1.env
conanfile.txt: Generator virtualenv created activate.sh
conanfile.txt: Generator virtualenv created deactivate.sh
conanfile.txt: Generator virtualenv created environment.sh.env
conanfile.txt: Generator 'CMakeDeps' calling 'generate()'
conanfile.txt: Aggregating env generators
conanfile.txt: Generated conaninfo.txt
conanfile.txt: Generated graphinfo

C:\Proj\pqtest\build>activate.bat

(conanenv) C:\Proj\pqtest\build>cmake .. -G "MinGW Makefiles" -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release
-- Using Conan toolchain: C:/Proj/pqtest/build/conan_toolchain.cmake
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Users/rainer/.conan/data/mingw-builds/12.2.0/_/_/package/6903a9d1b48b06f7fbe0929db654512c77e6cc32/bin/gcc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Users/rainer/.conan/data/mingw-builds/12.2.0/_/_/package/6903a9d1b48b06f7fbe0929db654512c77e6cc32/bin/g++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: Component target declared 'libpq::pgport'
-- Conan: Component target declared 'libpq::pgcommon'
-- Conan: Component target declared 'libpq::pq'
-- Conan: Target declared 'PostgreSQL::PostgreSQL'
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_POLICY_DEFAULT_CMP0091


-- Build files have been written to: C:/Proj/pqtest/build

(conanenv) C:\Proj\pqtest\build>mingw32-make
[ 50%] Building CXX object CMakeFiles/pqtest.dir/pqtest.cpp.obj
[100%] Linking CXX executable pqtest.exe
[100%] Built target pqtest

(conanenv) C:\Proj\pqtest\build>pqtest.exe    ---> cannot execute, libpq.dll not found
```

</details>

### PR DESCRIPTION
**Pull request by calebkiage**
Specify library name and version:  **lib/all**

Fixes #15301, Fixes #11716

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/libpq/all/conanfile.py [hunk 1]
```diff
@@ -189,6 +199,7 @@ def _remove_unused_libraries_from_package(self):
         else:
             globs = [
                 os.path.join(self.package_folder, "lib", "libpq.so*"),
+                os.path.join(self.package_folder, "lib", "*.dll"),
                 os.path.join(self.package_folder, "bin", "*.dll"),
                 os.path.join(self.package_folder, "lib", "libpq*.dylib")
             ]
```
Comment (SpaceIm) [inline recipes/libpq/all/conanfile.py:205]: I think cleanup is incomplete.
I propose to replace all the content of `_remove_unused_libraries_from_package()` by:

```python

    def _remove_unused_libraries_from_package(self):
        if self.settings.os == "Windows":
            rm(self, "*.dll", os.path.join(self.package_folder, "lib"))
            if self.options.shared:
                for lib in ("libpq.a", "libpgcommon.a", "libpgcommon_shlib.a", "libpgport.a", "libpgport_shlib.a"):
                    rm(self, lib, os.path.join(self.package_folder, "lib"))
            else:
                rm(self, "*.dll", os.path.join(self.package_folder, "bin"))
                rm(self, "*.dll.a", os.path.join(self.package_folder, "lib"))
        else:
            if self.options.shared:
                rm(self, "*.a", os.path.join(self.package_folder, "lib"))
            else:
                rm(self, "*.so*", os.path.join(self.package_folder, "lib"))
                rm(self, "*.dylib", os.path.join(self.package_folder, "lib"))

```

Comment (calebkiage) [reply recipes/libpq/all/conanfile.py:211]: I'll update my PR with changes from yours.

File: recipes/libpq/all/conanfile.py [hunk 2]
```diff
@@ -270,12 +264,16 @@ def package_info(self):
...
+                    if self.settings.os == "Windows":
+                        self.cpp_info.components["pgport"].system_libs = ["ws2_32"]
                     self.cpp_info.components["pgcommon"].requires.append("pgport")
 
         if self.settings.os in ["Linux", "FreeBSD"]:
             self.cpp_info.components["pq"].system_libs = ["pthread"]
         elif self.settings.os == "Windows":
             self.cpp_info.components["pq"].system_libs = ["ws2_32", "secur32", "advapi32", "shell32", "crypt32", "wldap32"]
+            if self.settings.get_safe("compiler.threads") == "posix":
+                self.cpp_info.components["pq"].system_libs.append("pthread")
```
Comment (SpaceIm) [inline recipes/libpq/all/conanfile.py:276]: it shouldn't be needed

Comment (calebkiage) [reply recipes/libpq/all/conanfile.py:276]: The library needs it, no? It calls functions from the threads lib

Comment (SpaceIm) [reply recipes/libpq/all/conanfile.py:276]: yes but explicit link is not required for MinGW

**Review by SpaceIm — APPROVED** [review]

**Review by jwillikers — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**